custom_model_runner/datarobot_drum/resource/transform_helpers.py

```diff
@@ -0,0 +1,40 @@
+import pyarrow as pa
+
+from io import BytesIO
+
+from scipy.io import mmwrite, mmread
+from scipy.sparse.csr import csr_matrix
+from scipy.sparse import vstack
+
+
+X_TRANSFORM_KEY = "X.transformed"
+
+
+def is_sparse(df):
+    return hasattr(df, "sparse") or type(df.iloc[0].values[0]) == csr_matrix
+
+
+def make_arrow_payload(df):
+    return pa.ipc.serialize_pandas(df, preserve_index=False).to_pybytes()
```

Comment:
have you tried deserializing this with python2.7?

Reply:
good call - will see how that works...presumably we particularly need to worry about the case where we're serializing in python3 and deserializing in python 2?

Reply:
@jlpolit for Arrow the version is payload
prior to Arrow v1.0.0 there're no backward/forward compatibility guarantees, so Arrow for instance if you serialized data with Arrow 0.24.1, you have to read it with 0.24.1

In the use cases where I added Arrow I check what Arrow version is installed in DRUM, and use Arrow to communicate between DRUM and DataRobot only if DRUM Arrow version works with DataRobot Arrow version (for Arrow < 1.0.0 that means exactly the same version)

Reply:
that tracks with what I'm seeing when trying to deserialize between drum and DR right now- what's your fallback behavior if the versions don't match up? just write_csv to bytesIO? sounds like I'll have to add that arrow version check here

Reply:
yes, the fallback is to use csv
I do this in 2 steps:
1. Check what Arrow version is on DRUM side (check DRUM's `/capabilities/` endpoint)
2. Decide if Arrow can be used. If not (versions are not compatible) - use csv

yes, you have to check that Arrow versions are the same before using Arrow

Reply:
so the use case you're talking about is whether you're going to pass an arrow object from DR to drum, right? which means if I want to determine whether to pass an arrow object the other direction I will need to either do that same check on the DR side and then pass, say, a `use_arrow` flag to the endpoint to tell drum what to do, or I'll have to pass my DR arrow version in the payload so drum can decide for itself

Reply:
> so the use case you're talking about is whether you're going to pass an arrow object from DR to drum, right?

yes!
this is prediction requests' data passed from DR to DRUM

> which means if I want to determine whether to pass an arrow object the other direction I will need to either do that same check on the DR side and then pass, say, a use_arrow flag to the endpoint to tell drum what to do, or I'll have to pass my DR arrow version in the payload so drum can decide for itself

yes
regarding your options, I like first one a little bit more
In DR, custom model capabilities (which include Arrow version) are already set to `CustomModelExecutable`. 
Then before doing predictions to CM model I check the capabilities and decide if I can use Arrow.
for your case, it may be easy to also check the capabilities and pass this `use_arrow` flag.
In this case we'll want have logic that checks Arrow versions compatibility duplicated in DR and DRUM.
Also note that the capabilities are not set to `CustomModelExecutable` for all execution paths yet (TBD in future). For instance, the capabilities aren't set to `CustomModelExecutable` for variable importance job.

Reply:
hey @rvorobii do you know if there's a way to tell pyarrow to serialize using an older version? I'm seeing some stuff in `pyarrow.ipc.IpcWriteOptions` that suggests it may be possible

Reply:
I'm not aware if this is possible.
Haven't seen anything related in Arrow docs.

Reply:
ok, I _think_ I have a solution

Reply:
@4dahalibut  the current implementation will take arrow version from DR, and if DR is using an older arrow than drum it returns in an older format- I've confirmed it works going from version 0.2 to 0.1.4. lmk if that seems like a reasonable solution to you (i.e. that it's not too onerous for DR to send an arrow version to drum when it calls this endpoint)

Reply:
I think that's fine

Reply:
@jlpolit also, we should be able to disable Arrow usage - it's not as battle-tested as csv is, so in case in there're any issues with Arrow we should be able to disable it. Or make it OFF by default, but let users to opt-in.

for instance, for my use case where DR can use Arrow to make predictions to DRUM, it's possible to disable Arrow via EngConfig, so that CSV is always used (this is especially important for On Prem releases).

Reply:
Roman - since our feature is only barely in alpha, its much more okay if it breaks.

Reply:
you guys know better

Reply:
as it stands right now, if you don't pass an arrow version to the endpoint it will still default to csv

Reply:
I think it should probably just fail but we can do it this way for now